packages/governance/src/typeGuards.js

```diff
@@ -0,0 +1,272 @@
+  YesApiInvocationPositionShape,
+  NoApiInvocationPositionShape,
+]);
+export const ApiInvocationQuestionSpecShape = harden({
+  method: 'unranked',
+  issue: ApiInvocationSpecShape,
+  positions: ApiInvocationPositionsShape,
+  electionType: 'api_invocation',
+  maxChoices: 1,
+  closingRule: ClosingRuleShape,
+  quorumRule: QuorumRuleShape,
+  tieOutcome: NoApiInvocationPositionShape,
+});
+export const ApiInvocationQuestionDetailsShape = harden({
+  method: 'unranked',
+  issue: ApiInvocationSpecShape,
+  positions: ApiInvocationPositionsShape,
+  electionType: 'api_invocation',
+  maxChoices: 1,
+  closingRule: ClosingRuleShape,
+  quorumRule: QuorumRuleShape,
+  tieOutcome: NoApiInvocationPositionShape,
+  questionHandle: makeHandleShape('Question'),
+  counterInstance: InstanceHandleShape,
+});
+
+const SimpleSpecShape = harden({
+  text: M.string(),
+});
+export const YesSimplePositionShape = harden({ text: M.string() });
+export const NoSimplePositionShape = harden({ text: M.string() });
+export const SimplePositionsShape = harden([
+  YesSimplePositionShape,
+  NoSimplePositionShape,
+]);
+export const SimpleIssueShape = SimpleSpecShape;
+export const SimpleQuestionSpecShape = harden({
+  method: ChoiceMethodShape,
+  issue: SimpleIssueShape,
+  positions: SimplePositionsShape,
+  electionType: M.or('election', 'survey'),
+  maxChoices: M.gte(1),
+  closingRule: ClosingRuleShape,
+  quorumRule: QuorumRuleShape,
+  tieOutcome: NoSimplePositionShape,
+});
+export const SimpleQuestionDetailsShape = harden({
+  method: ChoiceMethodShape,
+  issue: SimpleIssueShape,
+  positions: SimplePositionsShape,
+  electionType: M.or('election', 'survey'),
+  maxChoices: M.gte(1),
+  closingRule: ClosingRuleShape,
+  quorumRule: QuorumRuleShape,
+  tieOutcome: NoSimplePositionShape,
+  questionHandle: makeHandleShape('Question'),
+  counterInstance: InstanceHandleShape,
+});
+
+export const QuestionSpecShape = M.or(
+  ApiInvocationQuestionSpecShape,
+  OfferFilterQuestionSpecShape,
+  ParamChangesQuestionSpecShape,
+  SimpleQuestionSpecShape,
+);
+
+export const PositionShape = M.or(
+  YesApiInvocationPositionShape,
+  NoApiInvocationPositionShape,
+  YesOfferFilterPositionShape,
+  NoOfferFilterPositionShape,
+  YesSimplePositionShape,
+  NoSimplePositionShape,
+  YesParamChangesPositionShape,
+  NoParamChangesPositionShape,
+);
+
+export const QuestionHandleShape = makeHandleShape('question');
+
+// TODO(hibbert): add details; move to a more appropriate location
+export const SubscriberShape = M.remotable('Subscriber');
+export const InvitationShape = M.remotable('Invitation');
+
+// XXX I want to add questionHandle and counterInstance to
+// ParamChangesQuestionSpecShape. I don't see any alternative to adding the
+// methods to each member separately
+export const QuestionDetailsShape = M.or(
+  ParamChangesQuestionDetailsShape,
+  ApiInvocationQuestionDetailsShape,
+  OfferFilterQuestionDetailsShape,
+  SimpleQuestionDetailsShape,
+);
```

Comment:
Thanks for putting the question inline. Would be good for anything in common to be pulled out of the expensive `M.or`. If you pull these two common fields out of the branches, then you should also rename the branches to not imply that they're meaningful by themselves, and stop exporting them. For three of these, they are only used here, so no problem. But `ParamChangesQuestionDetailsShape` is used elsewhere. Assuming that this is dealt with somehow, we could rewrite the factored pattern as

```suggestion
export const QuestionDetailsShape = M.split({
    questionHandle: QuestionHandleShape, // making the handle shape should already be factored out.
    counterInstance: InstanceHandleShape,
  },
  M.or(
    ParamChangesQuestionPartialDetailsShape,
    ApiInvocationQuestionPartialDetailsShape,
    OfferFilterQuestionPartialDetailsShape,
    SimpleQuestionPartialDetailsShape,
  ),
});
```

Reply:
Thanks for the explanation. That's a helpful alternative.

Now that I see how `ParamChangesQuestionDetailsShape` is used, I wouldn't be surprised to find other uses of the other `QuestionDetailsShape`s. I think I'll leave it as is.